presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -261,13 +264,15 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
                     .collect(toImmutableList());
 
             List<List<Symbol>> inputs = new ArrayList<>();
+            Multimap<Symbol, Symbol> inputToOutputs = ArrayListMultimap.create();
```

Comment:
Instead of such multimap I would have a `Map<List<Symbol>, Symbol> exchangeNodeMapping` - a mapping of canonical inputs to canonical output. I think that way `getExchangeNodeOutputSymbolForInputSymbols(` will become much easier, as you will only canonicalize the input and check if it is are in this mappiing. No need for `outputAliases` variable and using `Set#retainAll` method.